feat: Implement Guest Mode for anonymous chatting

johannes117 opened the pull request:
This PR introduces support for guest users, allowing individuals to start conversations without signing in. It also adds a read-only mode for public chats.

**Key Changes:**

*   **Guest Session Management:** Unauthenticated users are assigned a `sessionId` (stored in local storage) to track their conversations and messages.
*   **Backend Support:** Convex functions and schema are updated to handle `sessionId` for guest conversations, including authorization checks.
*   **Guest Limitations:**
    *   Guests are limited to 10 messages per conversation. A prompt to sign in is displayed upon reaching the limit.
    *   File attachments are disabled for guest users.
*   **Data Cleanup on Login:** When a guest signs in, their temporary guest data (conversations and messages) is automatically cleared from the database.
*   **Read-Only/Viewer Mode:** Users viewing a public chat they don't own are now in a read-only mode and cannot send messages.
*   **Fix:** Corrected the URL generation for sharing public conversations to use the conversation's UUID.

johannes117 committed:
feat: Enhance guest session handling and attachment management

- Introduced sessionId support in attachment and conversation mutations, allowing guest users to save attachments and create conversations without authentication.
- Updated conversation and message queries to include sessionId, enabling guest access to their conversations and messages.
- Implemented a clearGuestData mutation to remove guest data upon user authentication, ensuring a seamless transition from guest to authenticated user.
- Enhanced ChatInput component to manage guest message limits and provide feedback for file attachments.
- Added a GuestMessageLimit component to inform users when they reach the message limit for guest accounts, prompting them to sign in for continued access.

johannes117 committed:
refactor: Simplify attachment management by removing sessionId and enforcing user authentication

- Removed optional sessionId from attachment saving logic, requiring user authentication for saving attachments.
- Updated schema to reflect the mandatory userId field for attachments, ensuring only authenticated users can upload files.
- Adjusted ChatInput component to restrict file uploads to authenticated users, enhancing security and user experience.

johannes117 committed:
fix: Update conversation access logic to handle public conversations

- Modified the `getById` query to return public conversations directly, enhancing access control.
- Updated the `MessageControls` component to generate share links based on the conversation's UUID, improving user experience when sharing public conversations.

johannes117 committed:
feat: Implement read-only mode for public conversations

- Added `isViewerMode` state to manage read-only access for users in public chats.
- Introduced `ReadOnlyPrompt` component to inform users they are in read-only mode.
- Updated `ChatInput` component to conditionally render the `ReadOnlyPrompt` based on user access level.
- Enhanced conversation logic to handle viewer mode appropriately, improving user experience in public conversations.

johannes117 committed:
chore: removed added fonts

SamDebono commented on convex/messages.ts:
```diff
@@ -192,29 +207,29 @@ export const list = query({
-        .withIndex("by_conversation_and_created", (q) => q.eq("conversationId", args.conversationId))
-        .order("asc")
-        .collect();
+      hasAccess = true;
```
Just looking at this, it could be better e.g.
`const isOwner = identity && conversation.userId === identity.subject;
const isGuestOwner = !identity && sessionId && conversation.sessionId === sessionId;
const isPublic = conversation.isPublic;

const hasAccess = isOwner || isGuestOwner || isPublic;
`

SamDebono commented on frontend/components/ChatInput.tsx:
```diff
@@ -178,56 +205,60 @@ function PureChatInput({ threadId, isStreaming, convexConversationId, onConvexCo
-      let attachments
-      if (currentFiles.length > 0) {
-        attachments = await uploadFiles(currentFiles)
+      const isNewThreadRoute = location.pathname === "/" || location.pathname === "/chat";
```
Dont we have constants for these that we can sub in?

johannes117 committed:
refactor: Simplify conversation access logic in `list` query

- Replaced multiple conditional checks for access control with a consolidated approach using boolean flags for owner, guest owner, and public status.
- Improved readability and maintainability of the access logic in the `list` query for conversations.
